### PR TITLE
支持在oi赛制中查看自己先前提交的代码

### PR DESCRIPTION
**Pull request by luanmenglei**

### CODE REVIEW

**Comment by luanmenglei** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by pandadtdyy** [comment]
you should remove record.ts which in wrong place

File: packages/hydrooj/src/handler/record.ts [hunk 1]
```diff
@@ -330,14 +356,15 @@ class RecordDetailConnectionHandler extends ConnectionHandler {
             rdoc.code = '';
             rdoc.compilerTexts = [];
         }
-
+
```
Comment (pandadtdyy) [inline packages/hydrooj/src/handler/record.ts]: maybe you need to run `yarn run lint`.

**Review by pandadtdyy — CHANGES_REQUESTED** [review]

File: packages/hydrooj/src/handler/record.ts [hunk 2]
```diff
@@ -152,6 +159,13 @@ class RecordDetailHandler extends ContestDetailBaseHandler {
             user.getById(domainId, rdoc.uid),
         ]);
 
+        if (contestSelfCode) {
+            rdoc.status = STATUS.STATUS_WAITING;
```
Comment (undefined-moe) [inline packages/hydrooj/src/handler/record.ts]: time? memory? message? etc.

File: packages/hydrooj/src/handler/record.ts [hunk 3]
```diff
@@ -134,14 +134,21 @@ class RecordDetailHandler extends ContestDetailBaseHandler {
     // eslint-disable-next-line consistent-return
     async get(domainId: string, rid: ObjectID, download = false) {
         const rdoc = this.rdoc;
+        let contestSelfCode = false;
```
Comment (undefined-moe) [inline packages/hydrooj/src/handler/record.ts]: Something called `canViewStatus` or `canViewDetail` might be better.

File: packages/hydrooj/src/handler/record.ts [hunk 4]
```diff
@@ -160,13 +174,23 @@ class RecordDetailHandler extends ContestDetailBaseHandler {
...
         } else if (download) return await this.download();
         if (pdoc && !(rdoc.contest && this.user._id === rdoc.uid)) {
-            if (!problem.canViewBy(pdoc, this.user)) throw new PermissionError(PERM.PERM_VIEW_PROBLEM_HIDDEN);
+            if (!problem.canViewBy(pdoc, this.user)) {
+                throw new PermissionError(PERM.PERM_VIEW_PROBLEM_HIDDEN);
+            } 
+        }
+
+        if (contestSelfCode) {
+            rdoc.status = STATUS.STATUS_WAITING;
```
Comment (undefined-moe) [inline packages/hydrooj/src/handler/record.ts]: Why appear twice?

File: packages/hydrooj/src/handler/record.ts [hunk 5]
```diff
@@ -308,36 +332,45 @@ class RecordDetailConnectionHandler extends ConnectionHandler {
 
     @param('rid', Types.ObjectID)
     async prepare(domainId: string, rid: ObjectID) {
+        let viewSelfCode = false;
```
Comment (undefined-moe) [inline packages/hydrooj/src/handler/record.ts]: Code-only views **should** never connect to this endpoint to receive updates, and **this should be done explicitly in frontend**.

**Review by undefined-moe — CHANGES_REQUESTED** [review]

File: packages/hydrooj/src/handler/record.ts (hunk 2 above)
Comment (undefined-moe) [reply packages/hydrooj/src/handler/record.ts]: What about simply pick the required properties instead of removing extra properties?

Comment (luanmenglei) [reply packages/hydrooj/src/handler/record.ts]: 直接重新弄一个可能会更麻烦

**Comment by undefined-moe** [comment]
Please solve errors in CI first

**Comment by undefined-moe** [comment]
By the way, I'm curious about how can you know the id for your submission in a contest before viewing it?

**Comment by luanmenglei** [comment]
> By the way, I'm curious about how can you know the id for your submission in a contest before viewing it?

image
clicking here

**Comment by undefined-moe** [comment]
So where is your link? I didn't see any modifications here

**Comment by luanmenglei** [comment]
> here

in fact, without this update, you can also click here, but the difference is that you will get a permission requiered message

**Comment by luanmenglei** [comment]
> > So where is your link? I didn't see any modifications here
in fact, without this update, you can also click here, but the difference is that you will get a permission requiered message

**Comment by luanmenglei** [comment]
> So where is your link? I didn't see any modifications here

in fact, without this update, you can also click here, but the difference is that you will get a permission requiered message

**Comment by undefined-moe** [comment]
You should goto `files` panel and revert changes to lots of unrelated lines.

**Comment by luanmenglei** [comment]
> You should goto `files` panel and revert changes to lots of unrelated lines.

Sorry but i can't actually understand what you mean, as i have poor knowledge about git

**Pull request closed without merging**